crates/ruff_python_semantic/src/model.rs

```diff
@@ -662,12 +662,104 @@ impl<'a> SemanticModel<'a> {
                 }
             }
 
+            // FIXME: Shouldn't this happen above where `class_variables_visible` is set?
             seen_function |= scope.kind.is_function();
         }
 
         None
     }
 
+    /// Simulates a runtime load of a given [`ast::ExprName`].
+    ///
+    /// This should not be run until after all the bindings have been visited.
+    pub fn simulate_runtime_load(&self, name: &ast::ExprName) -> Option<BindingId> {
+        let symbol = name.id.as_str();
+        let range = name.range;
+        let mut seen_function = false;
+        let mut class_variables_visible = true;
+        let mut lexicographical_lookup = true;
+        for (index, scope_id) in self.scopes.ancestor_ids(self.scope_id).enumerate() {
+            let scope = &self.scopes[scope_id];
+
+            // Only once we leave a function scope and its enclosing type scope should
+            // we stop doing lexicographical lookups. We could e.g. have nested classes
+            // where we lookup symbols from the innermost class scope, which can only see
+            // things from the outer class(es) that have been defined before the inner
+            // class.
+            if seen_function && !scope.kind.is_type() {
+                lexicographical_lookup = false;
+            }
+
+            if scope.kind.is_class() {
+                if seen_function && matches!(symbol, "__class__") {
+                    return None;
+                }
+                if !class_variables_visible {
+                    continue;
+                }
+            }
+
+            class_variables_visible = scope.kind.is_type() && index == 0;
+            seen_function |= scope.kind.is_function();
+
+            if let Some(binding_id) = scope.get(symbol) {
+                if lexicographical_lookup {
+                    // we need to look through all the shadowed bindings
+                    // since we may be shadowing a valid runtime binding
+                    // with an invalid one
```

Comment:
I _think_ that "valid" and "invalid" here mean "precedes the name node in source order" and "does not precede the name node in source order", but again I only figured that out with hindsight; this comment would have been more helpful if more explicit.